### PR TITLE
feat(tools): add HugeGraph AI DeepWiki assistant

### PR DESCRIPTION
**Pull request by LRriver**
## Purpose

Add an optional HugeGraph AI repository knowledge assistant under `tools/ai`. The assistant is intended for Claude Code and Codex users who want repository-scoped Q&A for https://github.com/apache/hugegraph-ai, using DeepWiki as the online knowledge source while caching wiki contents locally for repeated context search.

## Changes

- Add `tools/ai/hugegraph-ai-deepwiki-skill` as a standalone installable module.
- Include Claude and Codex plugin manifests plus marketplace manifests.
- Add bilingual installation and usage docs: `README.md` and `README-zh.md`.
- Add a small DeepWiki MCP client CLI for `structure`, `contents`, `context`, `ask`, and `tools`.
- Keep the assistant isolated from runtime code and project dependencies.

## Verification

- `python3 -m json.tool` on all new JSON manifests.
- `python3 -m py_compile tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py`.
- `uv run --extra dev ruff format --check tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py`.
- `uv run --extra dev ruff check tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py`.
- `claude plugin validate tools/ai/hugegraph-ai-deepwiki-skill`.
- `claude plugin validate tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill`.
- Temporary Codex install via `codex plugin marketplace add` and `codex plugin add`.
- DeepWiki smoke tests: `structure`, cached `context`, and online `ask` for `apache/hugegraph-ai`.

## Compatibility

This is an optional tool module only. It does not change HugeGraph AI runtime behavior, public APIs, package dependencies, or default configuration.

### CODE REVIEW

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    seen_payloads: list[str] = []
+    max_seconds = float(os.environ.get("DEEPWIKI_MCP_STREAM_TIMEOUT", "120"))
+    deadline = time.monotonic() + max_seconds
+    timed_out = False
+
+    while True:
+        if time.monotonic() > deadline:
+            timed_out = True
+            break
+        raw_line = response.readline()
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py]: Fixed in 3a3e8e6.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+            "Mcp-Protocol-Version": self.protocol_version,
+            "User-Agent": "hugegraph-ai-deepwiki-skill/0.1.4",
+        }
+        if self.session_id:
+            headers["Mcp-Session-Id"] = self.session_id
+
+        req = urllib.request.Request(self.endpoint, data=body, headers=headers, method="POST")
+        try:
+            with urllib.request.urlopen(req, timeout=90) as response:
+                session_id = response.headers.get("Mcp-Session-Id")
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:303]: Fixed in 3a3e8e6.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+
+def contents_cache_path(repo_name: str) -> Path:
+    return repo_cache_dir(repo_name) / "wiki-contents.md"
+
+
+def write_text_atomic(path: Path, text: str) -> None:
+    path.parent.mkdir(parents=True, exist_ok=True)
+    tmp_path = path.with_suffix(path.suffix + ".tmp")
+    tmp_path.write_text(text, encoding="utf-8")
+    tmp_path.replace(path)
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py]: Fixed in 3a3e8e6.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+        self.session_id: str | None = None
+        self.next_id = 1
+
+    def request(self, payload: dict[str, Any], expect_response: bool = True) -> dict[str, Any] | None:
+        body = json.dumps(payload).encode("utf-8")
+        headers = {
+            "Accept": "application/json, text/event-stream",
+            "Content-Type": "application/json",
+            "Mcp-Protocol-Version": self.protocol_version,
+            "User-Agent": "hugegraph-ai-deepwiki-skill/0.1.4",
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py]: Fixed in 3a3e8e6.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+            payload["params"] = params
+        self.request(payload, expect_response=False)
+
+    def initialize(self) -> None:
+        self.rpc(
+            "initialize",
+            {
+                "protocolVersion": self.protocol_version,
+                "capabilities": {},
+                "clientInfo": {"name": "hugegraph-ai-deepwiki-skill", "version": "0.1.4"},
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py]: Fixed in 3a3e8e6.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py]: Follow-up Python 3.9 socket timeout compatibility fix added in 8305e3d.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:303]: Follow-up Python 3.9 socket timeout compatibility fix added in 8305e3d.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+
+
+def ensure_cached_contents(client: McpClient, repo_name: str, refresh: bool = False) -> tuple[str, Path, bool]:
+    path = contents_cache_path(repo_name)
+    try:
+        if path.exists() and not refresh:
+            return path.read_text(encoding="utf-8"), path, False
+
+        text = read_wiki_contents(client, repo_name)
+        write_text_atomic(path, text)
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:404]: ‼️ **Do not discard freshly fetched contents on cache write failure**

Evidence: `read_wiki_contents()` runs before this write, but any `OSError` from `write_text_atomic()` is converted into `McpError`, so the command fails even though it already has the fresh wiki text. Impact: a broken cache directory makes `context`/`contents` unusable instead of falling back to the live result, which is the recovery path this change is trying to harden. Please return the fetched contents when only the cache write fails, and refetch when an existing cache read fails or is invalid.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+    try:
+        parsed = json.loads(data)
+    except json.JSONDecodeError as exc:
+        raise McpError(f"DeepWiki MCP returned non-JSON content: {preview_text(data)}") from exc
+    if not isinstance(parsed, dict):
+        raise McpError(f"DeepWiki MCP returned an unexpected JSON payload: {preview_text(data)}")
+    return parsed
+
+
+def read_sse_response(response: Any, expected_id: int | None) -> dict[str, Any]:
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:222]: ‼️ **Add regression tests for the MCP client failure paths**

Evidence: this PR adds a custom MCP/SSE client with timeout handling, atomic cache writes, CLI validation, and context scoring, but no automated tests cover those behaviors. Impact: the same timeout/cache paths already needed follow-up fixes in this PR, and future changes can regress them while `py_compile` and smoke checks still pass. Please add focused tests for `read_sse_response()`, cache read/write failure fallback, and the cached-context scoring/selection behavior.

**Review by imbajin — COMMENTED** [review]
Blocking: yes. Summary: The cache hardening still fails recoverable cache cases and the new MCP client lacks regression tests for its failure-prone paths. Evidence: py_compile/JSON smoke passed; fake-client cache write failure reproduced a hard McpError.

**Comment by LRriver** [comment]
> Blocking: yes. Summary: The cache hardening still fails recoverable cache cases and the new MCP client lacks regression tests for its failure-prone paths. Evidence: py_compile/JSON smoke passed; fake-client cache write failure reproduced a hard McpError.阻塞项：是。摘要：缓存加固仍无法处理可恢复的缓存故障场景，且新增的 MCP 客户端缺少针对其易故障路径的回归测试。证据：py_compile/JSON 冒烟测试通过；模拟客户端缓存写入失败复现了硬性 McpError 错误。

@imbajin Addressed the DeepWiki MCP review feedback in the latest `deepwiki-skill` commits. Summary:

- Cache read failures now trigger a fresh `read_wiki_contents()` fetch.
- Cache write failures no longer discard already fetched DeepWiki contents.
- Added regression tests for SSE timeout handling, cache fallback/refetch, and cached-context selection.
- Fixed Ruff format/lint issues from CI.

Local validation passed: `uv run ruff check .`, `uv run ruff format --check .`, `python3 -m py_compile ...`, and `python3 -m unittest .../tests/test_deepwiki_mcp.py`.

Latest `apache/hugegraph-ai#355` CI is passing.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 6 above)
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:404]: Fixed. `ensure_cached_contents()` now refetches when an existing cache read fails, and if `read_wiki_contents()` succeeds but the cache write fails, it returns the fresh DeepWiki contents instead of raising `McpError`. Added regression coverage for both cache write fallback and invalid cache refetch.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 7 above)
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:222]: Fixed. Added focused `unittest` coverage for `read_sse_response()` timeout handling, cache write fallback, invalid cache refetch, and cached-context scoring/selection. Local Ruff check/format, py_compile, and unit tests pass; the latest PR CI is also passing.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
+#!/usr/bin/env python3
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:1]: ‼️ **Core script is omitted from source archives**

Evidence: on current head `29aae41`, `git archive HEAD | tar -tf -` includes `tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/tests/test_deepwiki_mcp.py` but not this `scripts/deepwiki_mcp.py` file. The existing root `.gitattributes` has `scripts/ export-ignore`, so archive-based installs and release source tarballs ship the tests/manifests without the executable client. Please either move/rename the plugin script directory or add an explicit archive rule so the skill package is complete in source archives.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+            continue
+
+        data = "\n".join(data_lines)
+        data_lines = []
+        seen_payloads.append(data)
+        parsed = parse_json(data)
+        if expected_id is None or parsed.get("id") == expected_id:
+            return parsed
+
+    if data_lines:
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:261]: ⚠️ **Partial SSE timeouts still surface as parse errors**

Evidence: a fake response that first returns `b'data: {"jsonrpc":"2.0","id":7,\n'` and then raises `TimeoutError` makes `read_sse_response()` raise `McpError: DeepWiki MCP returned non-JSON content...` instead of the timeout error path added above. The new timeout test only covers an immediate timeout before any bytes are buffered. Please treat the buffered partial event as a timeout when `timed_out` is true before parsing the trailing `data_lines`, and add a regression test for that branch.

**Review by imbajin — COMMENTED** [review]
Blocking: yes. Summary: two packaging/runtime edge cases remain in the DeepWiki skill. Evidence: git archive/static review plus a local partial-SSE timeout reproduction.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 8 above)
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:1]: Fixed in the latest `deepwiki-skill` head. Added explicit archive rules in the root `.gitattributes` for this skill’s `scripts/` directory and files, and verified with `git archive HEAD | tar -tf -` that `scripts/deepwiki_mcp.py` is now included alongside the tests/manifests.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 9 above)
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:261]: Fixed in the latest `deepwiki-skill` head. `read_sse_response()` now treats buffered partial SSE data as a timeout when the stream times out instead of parsing incomplete JSON. Added `test_read_sse_response_reports_partial_event_timeout`; local Ruff, py_compile, and unittest checks pass.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+        raise McpError(f"Repository profile file could not be read: {REPOS_PATH}") from exc
+    except json.JSONDecodeError as exc:
+        raise McpError(f"Repository profile file is not valid JSON: {REPOS_PATH}") from exc
+
+    if not isinstance(repos, dict):
+        raise McpError(f"Repository profile file must contain a JSON object: {REPOS_PATH}")
+    return repos
+
+
+def resolve_repo(alias_or_name: str) -> str:
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:136]: ⚠️ **Accept the documented repository form**

Evidence: `resolve_repo("hugegraph-ai")` returns `apache/hugegraph-ai`, but `resolve_repo("apache/hugegraph-ai")` raises `McpError: Unknown repository alias ...`; the skill docs and metadata repeatedly present the canonical `apache/hugegraph-ai` value. Impact: users who copy the documented repository name into `--repo` hit a hard CLI error. Please either accept full `owner/repo` names as a pass-through or make the CLI/docs explicitly alias-only, and add a regression test for that contract.

**Review by imbajin — COMMENTED** [review]
Blocking: no. Summary: The skill package is mostly verified, but the CLI repository argument rejects a repository form the docs surface. Evidence: unittest passed (5 tests), py_compile passed, archive includes the skill script, and local resolve_repo("apache/hugegraph-ai") reproduction fails.

**Review by LRriver — COMMENTED** [review]
Submitting pending review draft before posting the latest follow-up replies.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py (hunk 10 above)
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:136]: Fixed in ad797fd. `resolve_repo()` now accepts both the documented canonical repository name `apache/hugegraph-ai` and the alias `hugegraph-ai`, while still limiting pass-through names to enabled repository profiles. Added `test_resolve_repo_accepts_alias_and_full_repo_name` for the contract.

File: tools/ai/hugegraph-ai-deepwiki-skill/README.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+- Answers repository-grounded questions about HugeGraph AI modules, agents, RAG workflows, graph-enhanced AI features, model integration, configuration, examples, installation, and implementation details.
+- Uses `read_wiki_contents` to build a local DeepWiki wiki cache and searches that cache before answering.
+- Uses `ask_question` when the cached context does not directly and precisely answer the question.
+- Avoids cloning upstream repositories for ordinary Q&A.
+
+## Requirements
+
+- Python 3.9 or later for the bundled MCP client script.
+- Network access to `https://mcp.deepwiki.com/mcp`.
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/README.md:24]: Fixed in ad797fd. The English README now states Python 3.10+ for the bundled MCP client, matching the repository `requires-python >=3.10` and Ruff target.

File: tools/ai/hugegraph-ai-deepwiki-skill/README-zh.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+- 回答 HugeGraph AI 模块、agent、RAG 工作流、图增强 AI、模型集成、配置、示例、安装和实现细节相关问题。
+- 使用 `read_wiki_contents` 构建本地 DeepWiki wiki 缓存，并优先搜索缓存。
+- 当缓存内容不能直接、精准回答问题时，使用 `ask_question` 获取线上答案。
+- 普通问答不会 clone 上游源码仓库。
+
+## 前置要求
+
+- Python 3.9 或更高版本，用于运行随附的 MCP 客户端脚本。
+- 当前环境需要能访问 `https://mcp.deepwiki.com/mcp`。
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/README-zh.md:24]: Fixed in ad797fd. The Chinese README now states Python 3.10+ for the bundled MCP client, matching this repository runtime requirement.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+# HugeGraph AI Repository Knowledge Assistant
+
+Answer questions about the Apache HugeGraph AI source repository. Use the official DeepWiki MCP server as the underlying knowledge retrieval channel.
+
+- Source repository: `https://github.com/apache/hugegraph-ai`
+- DeepWiki page: `https://deepwiki.com/apache/hugegraph-ai`
+- MCP endpoint: `https://mcp.deepwiki.com/mcp`
+- Default repository: `apache/hugegraph-ai`
+- Runtime requirements: Python 3.9+ and network access to the MCP endpoint.
```
[the remark this answers is not in the document]
Comment (LRriver) [reply tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/SKILL.md]: Fixed in ad797fd. `SKILL.md` now declares Python 3.10+ for the bundled MCP client.

File: .gitattributes [hunk 14]
```diff
@@ -14,5 +14,7 @@ hugegraph-ml/ export-ignore
 .idea/ export-ignore
 style/ export-ignore
 scripts/ export-ignore
+tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/ -export-ignore
+tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/** -export-ignore
```
[the remark this answers is not in the document]
Comment (LRriver) [reply .gitattributes:19]: Partially addressed in ad797fd. I re-tested `git archive` without these rules and the archive omitted `scripts/deepwiki_mcp.py` while keeping the tests, so the negation rules are required for source/archive installs. I kept them and added a comment explaining that they preserve the DeepWiki skill MCP client in release/source archives.

File: tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+                        raise McpError("DeepWiki MCP returned an empty response.")
+                    parsed = parse_json(text)
+        except urllib.error.HTTPError as exc:
+            details = exc.read().decode("utf-8", errors="replace")
+            content_type = exc.headers.get("Content-Type", "unknown")
+            raise McpError(f"DeepWiki MCP HTTP {exc.code} ({content_type}): {preview_text(details)}") from exc
+        except (TimeoutError, socket.timeout) as exc:  # noqa: UP041
+            raise McpError(f"DeepWiki MCP request timed out after {stream_timeout_seconds():.0f}s.") from exc
+        except urllib.error.URLError as exc:
+            if isinstance(exc.reason, (TimeoutError, socket.timeout)):
```
Comment (imbajin) [inline tools/ai/hugegraph-ai-deepwiki-skill/plugins/hugegraph-ai-deepwiki-skill/skills/hugegraph-ai-deepwiki-skill/scripts/deepwiki_mcp.py:323]: ⚠️ **Fix the current Ruff failure**

Evidence: `uv run ruff check tools/ai/hugegraph-ai-deepwiki-skill` fails on this head with `UP038` at this line. Impact: the bundled skill package cannot pass the repository's Ruff checks locally. Please use the Python 3.10 union form that the project lint rule expects.

```suggestion
            if isinstance(exc.reason, TimeoutError | socket.timeout):
```

**Review by imbajin — COMMENTED** [review]
Blocking: no. Summary: One current-head lint failure remains in the bundled DeepWiki client. Evidence: `uv run ruff check tools/ai/hugegraph-ai-deepwiki-skill` fails with UP038 at `deepwiki_mcp.py:323`.
